### PR TITLE
KAFKA-16243: Make sure that we do not exceed max poll interval inside poll

### PR DESCRIPTION
**Pull request by lucasbru**
The consumer keeps a poll timer, which is used to ensure liveness of the application thread. The poll timer automatically updates while the `Consumer.poll(Duration)` method is blocked, while the newer consumer only updates the poll timer when a new call to `Consumer.poll(Duration)` is issued. This means that the kafka-console-consumer.sh tools, which uses a very long timeout by default, works differently with the new consumer, with the consumer proactively rejoining the group during long poll timeouts.

This change solves the problem by (a) repeatedly sending `PollApplicationEvents` to the background thread, not just on the first call of `poll` and (b) making sure that the application thread doesn't block for so long that it runs out of `max.poll.interval`.

An integration test is added to make sure that we do not rejoin the group when a long poll timeout is used with a low `max.poll.interval`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by AndrewJSchofield — APPROVED** [review]
lgtm

**Comment by lucasbru** [comment]
@cadonna could you please have a look?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (lianetm) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: nit: semicolon not needed here, same in the next 2 lines

**Comment by lianetm** [comment]
thanks for the changes! lgtm.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (lucasbru) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Done

**Comment by lianetm** [comment]
I like the last commit msg :)

**Comment by lucasbru** [comment]
Hey @mjsax . This PR is right now waiting for @cadonna's review. If you have time, you could take a look to get into the 848 work.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: nit:
This comment is a bit confusing. What is it supposed to clarify?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Do we need this comment? I think it would be better to delete it and to rename `initialAssignedCalls` to something more meaningful like `callsToAssignedAfterFirstRebalance` or `callsToAssignedBeforePolls`.

**Review by cadonna — APPROVED** [review]
Thanks for the PR, @lucasbru !

I have two minor comments about comments 🙂 .

For the rest, LGTM!

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (lucasbru) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Done

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (lucasbru) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Done

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:250]: For my own education: why `/ 2` ?

Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:250]: It's somewhat arbitrary. We need to make sure that the application thread doesn't block so long that the poll timer expires. We want to let it unblock sometime before the timer expires, and send an event to the background thread that it's still polling, and give the background thread time to process the event and reset the poll timer.

**Pull request merged**
